# Basic EDA


In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [2]:
path = '/content/gdrive/My Drive/Flipr 6//'

In [3]:
import pandas as pd
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

import numpy as np

def mean_absolute_percentage_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100


In [4]:
df = pd.read_excel(path+'Test-dataset.xlsx', sheet_name="Foreign_Visitors_TS")

In [5]:
df  =  df.iloc[1:]
df.columns  =  ['City', 'April', 'May', 'June', 'July', 'August']
from sklearn.utils import shuffle
df  =  shuffle(df)

In [6]:
df.head()

City   April     May    June    July  August
94      Sarangpur     NaN  294964  358178  379237  421365
137     Sitarganj   61420   74125   90018   95302  105882
279     Lohardaga   97323  117458  142635  151015  167785
100       Patratu   97323  117458  142635  151015  167785
454  Bagha Kusmar  535775  646624  785194  831371  923737

# Relation between the columns?

#### Using PyCaret for quick prediction.

In [7]:
del df['City']
display(df.shape)
df = df.dropna()
display(df.shape)

(501, 5)

(407, 5)

In [8]:
try:
    from pycaret.regression import *
except:
    !pip install pycaret==2.0
    from pycaret.regression import *

In [9]:
# Train Test Split.
train_, test_  =  df.iloc[:int(0.8*len(df))], df.iloc[int(0.8*len(df)):]
df.shape, train_.shape, test_.shape

((407, 5), (325, 5), (82, 5))

In [10]:
train_.head()

April      May     June     July   August
137    61420    74125    90018    95302   105882
279    97323   117458   142635   151015   167785
100    97323   117458   142635   151015   167785
454   535775   646624   785194   831371   923737
231  2717138  3279303  3982019  4216244  4684707

In [11]:
reg  =  setup(train_, 
              target = 'August')

 
Setup Succesfully Completed.


In [12]:
best_model = compare_models(fold=5)

Since even the simplest models show almost perfect score there is a strong relation, digging a little deeper we find the following - 

In [13]:
# df['May']     =  df['April']*1.2068969800369 - 2.612513869 + c
# df['June']    =  df['May']  *1.2142848513404 + 7.683109226 + c
# df['July']    =  df['June'] *1.0588299671675 - 12.5999801  + c
# df['August']  =  df['July'] *1.1111038027494 - 6.672843276 + c

# where c is a very small constant.


# BELOW IS THE VALIDATION - MAPE is almost 0.
print(  mean_absolute_percentage_error(df['April']*1.2068969800369 -2.612513869, df['May']), \
        mean_absolute_percentage_error(df['May']  *1.2142848513404 + 7.683109226, df['June']), \
        mean_absolute_percentage_error(df['June'] *1.0588299671675 - 12.5999801,  df['July']), \
        mean_absolute_percentage_error(df['July'] *1.1111038027494 - 6.672843276, df['August'])  )


0.0016124273940893684 0.00101758632382994 0.0022663719934442462 0.002861744639518736


### As constants are negligible compared to the absolute values, let's ignore and focus only on the '% change'
### From April to May:     +20.7%
### From May to June:      +21.4%
### From June to July:     +05.9%
### From July to August:   +11.1%

### Essentially, we have only 5 data points in the dataset, so we calculate 
### '% Change From August to September' = Average of Median + Mean

In [15]:
percentage_change_list = np.array([0.2068969800369, 0.2142848513404, 0.0588299671675, 0.1111038027494])
percentage_change = (np.mean(percentage_change_list)  +  np.median(percentage_change_list))  /  2
percentage_change

0.15338964585835002

### So, Sepetember will have an increase in cases from August by 15.3%. <br> [And, we use this to predict in 01.ipynb]